server/core/storelimit/sliding_window.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2022 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storelimit
+
+import "github.com/tikv/pd/pkg/syncutil"
+
+const (
+	snapSize = 10
+)
+
+// SlidingWindows is a multi sliding windows
+type SlidingWindows struct {
+	mu      syncutil.RWMutex
+	windows []*window
+}
+
+// NewSlidingWindows is the construct of SlidingWindows.
+func NewSlidingWindows(capacity int64) *SlidingWindows {
+	windows := make([]*window, MaxPriority)
+	for i := 0; i < int(MaxPriority); i++ {
+		windows[i] = newWindow(capacity >> i)
+	}
+
+	return &SlidingWindows{
+		windows: windows,
+	}
+}
+
+// Reset resets the capacity of the sliding windows.
+func (s *SlidingWindows) Reset(capacity int64) {
+	if s == nil {
+		return
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	for i, v := range s.windows {
+		v.reset(capacity >> i)
+	}
+}
+
+// GetUsed returns the used size in the sliding windows.
+func (s *SlidingWindows) GetUsed() int64 {
+	if s == nil {
+		return 0
+	}
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	used := int64(0)
+	for _, v := range s.windows {
+		used += v.getUsed()
+	}
+	return used
+}
+
+// Available returns whether the token can be taken.
+// It will check the given window finally if the lower window has no free size.
+func (s *SlidingWindows) Available(_ int64, level Level) bool {
+	if s == nil {
+		return true
+	}
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	for i := 0; i < int(level); i++ {
+		if s.windows[i].available() {
+			return true
+		}
+	}
+	return false
+}
+
+// Take tries to take the token.
+// It will consume the given window finally if the lower window has no free size.
+func (s *SlidingWindows) Take(token int64, level Level) bool {
+	if s == nil {
+		return true
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	for i := 0; i <= int(level); i++ {
+		if s.windows[i].take(token) {
+			return true
+		}
+	}
+	return false
+}
+
+// Ack indicates that some executing operator has been finished.
+// It will refill the highest window first.
+func (s *SlidingWindows) Ack(token int64) {
+	if s == nil {
+		return
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	for i := MaxPriority - 1; i >= 0; i-- {
+		if token = s.windows[i].ack(token); token <= 0 {
+			break
+		}
+	}
+}
+
+// window is a sliding window.
+type window struct {
+	capacity int64
+	used     int64
+	count    uint
```

Comment:
What is the purpose of the field?

Reply:
// capacity is the max size of the window.
	capacity int64
	// used is the size of the operators that are executing.
	// the used can be larger than the capacity.but allow
	// one operator to exceed the capacity at most.
	used int64
	// the count of operators in the window.
	count uint

Reply:
Yes, but I don't find anywhere use it in this PR.

Reply:
It is used by the later PR: https://github.com/bufferflies/pd/pull/8/files